### PR TITLE
remove or encapsulate the remaining non-query data in tcx

### PR DESCRIPTION
**Pull request by nikomatsakis**
I wound up removing the existing cache around inhabitedness since it didn't seem to be adding much value. I reworked const rvalue promotion, but not that much (i.e., I did not split the computation into bits, as @eddyb had tossed out as a suggestion). But it's now demand driven, at least.

cc @michaelwoerister -- see the `forbid_reads` change in last commit

r? @eddyb -- since the trickiest of this PR is the work on const rvalue promotion

cc #44137

### CODE REVIEW

**Comment by eddyb** [comment]
r=me modulo the forbid_reads hack. Also, cc @arielb1

**Comment by michaelwoerister** [comment]
`forbid_reads` is considered a hack?

**Comment by eddyb** [comment]
@michaelwoerister I meant `RegionEraser` and how it uses `forbid_reads` - to me it looks hackier than using a query, and it's only for performance reasons? But I'll defer that to you and/or @arielb1.

**Comment by Mark-Simulacrum** [comment]
Before we merge, let's get some perf numbers - run a try build when ready.

**Comment by nikomatsakis** [comment]
@eddyb I think that example could indeed be a query -- I was considering actually extending the maps macro to support such queries (i.e., queries that are not permitting to read during their code, and hence don't require a task) or just making it *be* a query. I could go either way.

File: src/librustc/ty/erase_regions.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+use rustc_data_structures::fx::FxHashMap;
```
Comment (kennytm) [inline src/librustc/ty/erase_regions.rs]: Missing license. 

```
[00:02:50] tidy error: /checkout/src/librustc/ty/erase_regions.rs: incorrect license
```

**Comment by arielb1** [comment]
LGTM

**Comment by kennytm** [comment]
Could not compile stage1-std in `x86_64-gnu-incremental` with a `BorrowMutError` somewhere.

```
[00:18:53] error: internal compiler error: unexpected panic
[00:18:53] 
[00:18:53] note: the compiler unexpectedly panicked. this is a bug.
[00:18:53] 
[00:18:53] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[00:18:53] 
[00:18:53] note: rustc 1.22.0-dev running on x86_64-unknown-linux-gnu
[00:18:53] 
[00:18:53] thread 'rustc' panicked at 'already borrowed: BorrowMutError', /checkout/src/libcore/result.rs:906:4
[00:18:53] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:18:53] 
[00:18:53] error: Could not compile `std`.
```

**Comment by alexcrichton** [comment]
@bors: r-

reappeared in the travis queue

**Comment by aidanhs** [comment]
```
[00:49:06] [1m[31merror[E0624](B[m[1m: method `uninhabited_from` is private(B[m
[00:49:06]     (B[m[1m[34m--> (B[m/checkout/src/tools/miri/src/librustc_mir/interpret/eval_context.rs:2241:8(B[m
[00:49:06]      (B[m[1m[34m|(B[m
[00:49:06] [1m[34m2241(B[m (B[m[1m[34m| (B[m    ty.uninhabited_from(&mut HashMap::default(), tcx).is_empty()(B[m
[00:49:06]      (B[m[1m[34m| (B[m       (B[m[1m[31m^^^^^^^^^^^^^^^^(B[m
[00:49:06] 
[00:49:07] [1m[31merror(B[m[1m: aborting due to previous error(B[m
[00:49:07] 
```

**Comment by alexcrichton** [comment]
@nikomatsakis I think you'll want to edit `src/tools/toolstate.toml`

**Comment by nikomatsakis** [comment]
OK, I *think* I followed the directions from CONTRIBUTING.md correctly. solson/miri#349 will fix this problem.

File: src/librustc/dep_graph/raii.rs [hunk 2]
```diff
@@ -55,3 +55,4 @@ impl<'graph> Drop for IgnoreTask<'graph> {
     }
 }
 
+
```
Comment (bjorn3) [inline src/librustc/dep_graph/raii.rs]: Why is this extra newline?

File: src/librustc/middle/mem_categorization.rs [hunk 3]
```diff
@@ -392,21 +394,46 @@ impl MutabilityCategory {
...
+    /// - similarly, as the results of upvar analysis are not yet
+    ///   known, the results around upvar accesses may be incorrect.
     pub fn with_infer(infcx: &'a InferCtxt<'a, 'gcx, 'tcx>,
                       region_scope_tree: &'a region::ScopeTree,
                       tables: &'a ty::TypeckTables<'tcx>)
                       -> MemCategorizationContext<'a, 'gcx, 'tcx> {
+        let tcx = infcx.tcx;
+
+        // Subtle: we can't do rvalue promotion analysis until the
+        // typeck false is complete, which means that you can't trust
```
Comment (bjorn3) [inline src/librustc/middle/mem_categorization.rs]: *phase?

**Review by bjorn3 — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
@bjorn3 thanks

**Comment by eddyb** [comment]
@nikomatsakis You can just not change `miri` now, thanks to https://github.com/solson/miri/pull/350.

**Comment by nikomatsakis** [comment]
@eddyb ok I dropped that commit

**Comment by eddyb** [comment]
r=me unless appveyor failure is a real problem you've yet to fix.

**Comment by nikomatsakis** [comment]
Argh, I thought this had landed. =)

**Comment by kennytm** [comment]
Still failed to build stage1-std on incremental. Same error message as https://github.com/rust-lang/rust/pull/44501#issuecomment-330155424.

**Comment by nikomatsakis** [comment]
!@#!$! I thought I fixed that... maybe not.

**Comment by nikomatsakis** [comment]
Hmm. I don't see this locally.

**Comment by kennytm** [comment]
ICE when building `stage1-std` in the incremental job.

<details>

```
[00:19:12] error: internal compiler error: /checkout/src/librustc/dep_graph/graph.rs:842: dep-graph reads are forbidden here, but DepNodeIndex { index: 26532 } was read
[00:19:12] 
[00:19:12] note: the compiler unexpectedly panicked. this is a bug.
[00:19:12] 
[00:19:12] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[00:19:12] 
[00:19:12] note: rustc 1.22.0-dev running on x86_64-unknown-linux-gnu
[00:19:12] 
[00:19:12] thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:492:8
[00:19:12] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:19:12] 
[00:19:12] error: Could not compile `std`.
```

</details>

**Comment by nikomatsakis** [comment]
Huh. I thought this was now working. Will have to investigate tomorrow in more depth.

**Comment by nikomatsakis** [comment]
Well, even though I can't reproduce the problem, it seems clearly to be about the `forbid_edges` code. Reading more carefully, I can *imagine* how a read could occur in that bit of code, so I just converted the relevant stuff to a query. 

r? @michaelwoerister on the last few commits

**Comment by eddyb** [comment]
@nikomatsakis Can you squash away the `forbid_reads` logic?

**Comment by nikomatsakis** [comment]
@eddyb done

**Comment by arielb1** [comment]
@nikomatsakis 

What can `erase_regions` depend on?

File: src/librustc/dep_graph/raii.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc/dep_graph/raii.rs]: Hehe it's still here.

**Comment by kennytm** [comment]
<details><summary>Clippy needs to be broken.</summary>

```
[00:36:46] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[00:36:46]   --> /checkout/src/tools/clippy/clippy_lints/src/escape.rs:74:29
[00:36:46]    |
[00:36:46] 74 |         ExprUseVisitor::new(&mut v, cx.tcx, cx.param_env, region_scope_tree, cx.tables).consume_body(body);
[00:36:46]    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 6 parameters
[00:36:46] 
[00:36:46] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[00:36:46]   --> /checkout/src/tools/clippy/clippy_lints/src/escape.rs:74:29
[00:36:46]    |
[00:36:46] 74 |         ExprUseVisitor::new(&mut v, cx.tcx, cx.param_env, region_scope_tree, cx.tables).consume_body(body);
[00:36:46]    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 6 parameters
[00:36:46] 
[00:36:46] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[00:36:46]     --> /checkout/src/tools/clippy/clippy_lints/src/loops.rs:1420:25
[00:36:46]      |
[00:36:46] 1420 |     ExprUseVisitor::new(&mut delegate, cx.tcx, cx.param_env, region_scope_tree, cx.tables).walk_expr(body);
[00:36:46]      |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 6 parameters
[00:36:46] 
[00:36:46] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[00:36:46]     --> /checkout/src/tools/clippy/clippy_lints/src/loops.rs:1420:25
[00:36:46]      |
[00:36:46] 1420 |     ExprUseVisitor::new(&mut delegate, cx.tcx, cx.param_env, region_scope_tree, cx.tables).walk_expr(body);
[00:36:46]      |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 6 parameters
[00:36:46] 
[00:36:47] error[E0609]: no field `rvalue_promotable_to_static` on type `rustc::ty::TyCtxt<'_, '_, '_>`
[00:36:47]    --> /checkout/src/tools/clippy/clippy_lints/src/methods.rs:806:14
[00:36:47]     |
[00:36:47] 806 |             .rvalue_promotable_to_static
[00:36:47]     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^ unknown field
[00:36:47] 
[00:36:47] error[E0609]: no field `rvalue_promotable_to_static` on type `rustc::ty::TyCtxt<'_, '_, '_>`
[00:36:47]    --> /checkout/src/tools/clippy/clippy_lints/src/methods.rs:806:14
[00:36:47]     |
[00:36:47] 806 |             .rvalue_promotable_to_static
[00:36:47]     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^ unknown field
[00:36:47] 
[00:36:47] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[00:36:47]    --> /checkout/src/tools/clippy/clippy_lints/src/needless_pass_by_value.rs:111:38
[00:36:47]     |
[00:36:47] 111 |             euv::ExprUseVisitor::new(&mut ctx, cx.tcx, cx.param_env, region_scope_tree, cx.tables).consume_body(body);
[00:36:47]     |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 6 parameters
[00:36:47] 
[00:36:47] error[E0061]: this function takes 6 parameters but 5 parameters were supplied
[00:36:47]    --> /checkout/src/tools/clippy/clippy_lints/src/needless_pass_by_value.rs:111:38
[00:36:47]     |
[00:36:47] 111 |             euv::ExprUseVisitor::new(&mut ctx, cx.tcx, cx.param_env, region_scope_tree, cx.tables).consume_body(body);
[00:36:47]     |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 6 parameters
[00:36:47] 
[00:36:48] error: aborting due to 4 previous errors
[00:36:48] 
[00:36:48] error: Could not compile `clippy_lints`.
[00:36:48] warning: build failed, waiting for other jobs to finish...
[00:36:48] error: aborting due to 4 previous errors
[00:36:48] 
[00:36:48] error: Could not compile `clippy_lints`.
[00:36:48] warning: build failed, waiting for other jobs to finish...
[00:36:59] error: build failed
```

</details>

**Comment by nikomatsakis** [comment]
cc @Manishearth @llogiq @mcarton @oli-obk -- apparently [I've broken clippy](https://github.com/rust-lang/rust/pull/44501#issuecomment-336704522). Looks like primarily this is a matter of now [needing to supply the `rvalue_promotable_map` to the `ExprUseVisitor`](https://github.com/nikomatsakis/rust/blob/dd3a4f45710163ed91efe582c5d0d7b4402d4f75/src/librustc/middle/expr_use_visitor.rs#L272-L274), but also the removal of the `rvalue_promotable_to_static` map (replaced by the `rvalue_promotable` query, [example edit](https://github.com/rust-lang/rust/pull/44501/files#diff-807bbbef86b78462b44b1f4399d9239dL874)). 

(Are there some directions for the steps to prepare a PR to clippy to fix this stuff?)

**Comment by oli-obk** [comment]
> (Are there some directions for the steps to prepare a PR to clippy to fix this stuff?)

Directions can be found at https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#external-dependencies

Generally, you can just fix clippy by compiling it via `./x.py build src/tools/clippy` and open a PR once it builds. We'll take care to reenable clippy once your changes hit master and are merged into clippy master.

Remember to set `submodules = false` in `config.toml`, otherwise the next call to `x.py` will erase your submodule changes

If you consider the PR to clippy to be too involved (e.g. because we used an API so wrongly it'll take a lot of effort to fix our usage), we'll fix it ourselvels.

File: src/librustc/dep_graph/raii.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc/dep_graph/raii.rs]: What's a newline between friends?

**Comment by nikomatsakis** [comment]
@arielb1 

> What can `erase_regions` depend on?

Actually, looking again, I'm not sure. I thought that I had seen -- or maybe just imagined -- some path in `anonymize_late_bound_regions` that accesses one of the adt-defs or something, but looking again I don't see any such path. Perhaps there was some other bug in `forbid_reads` the way I wrote it? I wish I could reproduce the problem.

Regardless, as @eddyb pointed out earlier, a regular query is simpler, so that may have been a bit of premature perf. optimization on my part. I think my ideal preference would be to keep it as a query, but add an option to declare a "non-read" query that skips dependency tracking (using something like `forbid_reads` to check for correctness).

**Comment by kennytm** [comment]
@bors retry

macOS... spuriously canceled after 6 hours 😂

**Comment by oli-obk** [comment]
Would it make sense to simplify the map returned by `rvalue_promotable_map` to a set?

**Pull request merged**